File: Java/LinkedList/LinkedList.java
import java.util.Arrays;

public class LinkedList {
  private Node head = null;
  private Node tail = null;
  private int size = 0;

  public void prepend(int value) {
    size++;
    Node newNode = new Node(value);
    if (head == null) {
      head = newNode;
      tail = newNode;
    } else {
      newNode.setNext(head);
      head = newNode;
    }
  }

  public void append(int value) {
    size++;
    Node newNode = new Node(value);
    if (tail == null) {
      head = newNode;
      tail = newNode;
    } else {
      tail.setNext(newNode);
      tail = newNode;
      tail.setNext(null);
    }
  }

  public void insert(int index, int value) {
    size++;
    Node newNode = new Node(value);
    if (index == 0) {
      prepend(value);
    } else if (index > size) {
      throw new ArrayIndexOutOfBoundsException("Out of range!");
    } else {
      int i = 0;
      Node prevNode = head;
      while (i < index - 1) {
        prevNode = prevNode.getNext();
        i++;
      }
      if (prevNode.getNext() == null) {
        append(value);
      }
      newNode.setNext((prevNode.getNext()));
      prevNode.setNext(newNode);
    }
  }

  public void replace(int index, int value) {
    int i = 0;
    Node newNode = head;
    while (i < index) {
      newNode = newNode.getNext();
      i++;
    }
    newNode.setNode(value);
  }

  public int delete(int index) {
    size--;
    if (index == 0) {
      Node output = head;
      head = head.getNext();
      return output.getNode();
    } else if (index > size) {
      throw new ArrayIndexOutOfBoundsException("Out of range!");
    } else {
      Node prevNode = head;
      int i = 0;
      while (i < index - 1) {
        prevNode = head.getNext();
        i++;
      }
      Node currentNode = prevNode.getNext();
      if (currentNode.getNext() == null) {
        prevNode.setNext(null);
        return currentNode.getNode();
      } else {
        Node deleteNode = currentNode.getNext();
        prevNode.setNext(deleteNode);
        Node output = currentNode.getNext();
        return output.getNode();
      }
    }
  }

  public int getIndex(int index) {
    if (index == 0) {
      return head.getNode();
    } else {
      int i = 0;
      Node currentNode = head;
      while (i != index) {
        currentNode = currentNode.getNext();
        i++;
      }
      return currentNode.getNode();
    }
  }

  public String getList() {
    int[] output = new int[size];
    int i = 0;
    Node currentNode = head;
    while (i < size) {
      output[i] = currentNode.getNode();
      currentNode = currentNode.getNext();
      i++;
    }
    return Arrays.toString(output);
  }

  public int getSize() {
    return size;
  }

  public String reverse() {
    Node prev_node = head;
    head = tail;
    tail = prev_node;
    if (size > 2) {
      Node current_node = prev_node.getNext();
      Node next_node = current_node.getNext();
      while (next_node != null) {
        current_node.setNext(prev_node);
        prev_node = current_node;
        current_node = next_node;
        next_node = next_node.getNext();
      }
      tail.setNext(null);
    }
    head.setNext(prev_node);
    return getList();
  }
}

File: JavaScript/LinkedList.js
class Node {
  constructor(value) {
    this.value = value;
    this.nextNode = null;
  }

  getValue() {
    return this.value;
  }

  setValue(value) {
    this.value = value;
  }

  getNext() {
    return this.nextNode;
  }

  setNext(nextNode) {
    this.nextNode = nextNode;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  append(value) {
    const newNode = new Node(value);
    this.length++;
    if (this.tail == null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.setNext(newNode);
      this.tail = newNode;
    }
  }
  prepend(value) {
    this.length++;
    const newNode = new Node(value);
    if (this.head == null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.setNext(this.head);
      this.head = newNode;
    }
  }

  insert(index, value) {
    if (index > this.length || index < 0) {
      throw "Index out of range";
    } else if (index == 0) {
      this.prepend(value);
    } else if (index == this.length) {
      this.append(value);
    } else {
      this.length++;
      const newNode = new Node(value);
      var prevNode = this.head;
      var currentNode = this.head.getNext();

      for (var i = 0; i < index - 1; i++) {
        prevNode = currentNode;
        currentNode = currentNode.getNext();
      }
      prevNode.setNext(newNode);
      newNode.setNext(currentNode);
    }
  }

  delete(index) {
    if (index > this.length || index < 0) {
      throw "Index out of range";
    }
    var delNode;
    this.length--;
    if (index == 0) {
      delNode = this.head;
      this.head = this.head.getNext();
      return delNode;
    } else {
      var prevNode = this.head;
      var currentNode = this.head.getNext();
      for (var i = 0; i < index - 1; i++) {
        prevNode = currentNode;
        currentNode = currentNode.getNext();
      }
      delNode = currentNode;
      prevNode.setNext(currentNode.getNext());
      return delNode;
    }
  }
  getList() {
    if (this.length == 0) {
      return [];
    } else {
      var currentNode = this.head;
      var nextNode = this.head.getNext();
      var list = [];
      for (var i = 0; i < this.length - 1; i++) {
        list.push(currentNode.getValue());
        currentNode = nextNode;
        nextNode = nextNode.getNext();
      }
      list.push(currentNode.getValue());
      return list;
    }
  }

  reverse() {
    const oldTail = this.tail;
    const oldHead = this.head;
    if (this.length <= 1) {
    } else if (this.length == 2) {
      this.head = oldTail;
      this.tail = oldHead;
    } else {
      var prevNode = this.head;
      var currentNode = this.head.getNext();
      var nextNode = currentNode.getNext();
      for (var i = 0; i < this.length - 2; i++) {
        currentNode.setNext(prevNode);
        prevNode = currentNode;
        currentNode = nextNode;
        nextNode = nextNode.getNext();
      }
      currentNode.setNext(prevNode);
      this.head = oldTail;
      this.tail = oldHead;
    }
    return this.getList();
  }
}

var newList = new LinkedList();

newList.append(4);
newList.append(5);
newList.prepend(3);
newList.insert(2, 2);
console.log(newList.getList());
console.log(newList.reverse());
